File: core/src/algorithms/sorting.cpp
#include "algorithms/sorting.h"
#include <vector>
#include <iostream>

Sorting::Sorting(std::vector<int> a){
    data = a;
    n = data.size();
    no_recursive_calls = 0;
    no_timesteps = 0;
    no_swaps = 0;
}
void Sorting::selection(){
    no_timesteps = 0;
    no_swaps = 0;
    for (int i=0; i<n-1; i++){
        int minIdx = i;
        for (int j = i+1; j<n; j++){
            if (data[j]<data[minIdx])
                minIdx = j;
            no_timesteps += 1;
        }
        std::swap(data[minIdx],data[i]);
        no_swaps += 1;
    }
}

void Sorting::insertion(){
    no_timesteps = 0;
    no_recursive_calls = 0;
    no_swaps = 0;
    bool inside;
    for (int i=1; i<n; i++){
        int current = data[i];
        int j = i-1;
        no_timesteps += 1;
        inside = false;
        while(j>=0 && data[j]>current){
            inside = true;
            no_timesteps += 1;
            data[j+1] = data[j];
            j = j-1;
        }
        if (inside){
            no_timesteps -= 1;
        }
        data[j+1] = current;
    }
}

void Sorting::bubble(){
    no_recursive_calls = 0;
    no_swaps = 0;
    no_timesteps = 0;
    for (int i=0; i<n-1; i++){
        bool swapped = false;
        for (int j=0; j<n-i-1; j++){
            no_timesteps += 1;
            if (data[j]>data[j+1]) {
                std::swap(data[j],data[j+1]);
                swapped = true;
                no_swaps += 1;
            }
        }
        if (!swapped){
            break;
        }
    }
}
void Sorting::merge(){
    no_recursive_calls = 0;
    no_swaps = 0;
    no_timesteps = 0;
    merge_sort(data, 0, n-1);
}

void Sorting::merge_sort(std::vector<int>& dt, int lower, int upper){
    no_recursive_calls += 1;
    if (lower >= upper)
        return;
    int mid = lower + (upper-lower)/2;
    merge_sort(dt, lower, mid);
    merge_sort(dt, mid+1, upper);
    int n1 = mid - lower + 1;
    int n2 = upper - mid;

    std::vector<int> left(n1);
    std::vector<int> right(n2);

    for (int i=0; i<n1; i++){
        left[i] =(dt[lower+i]);
        no_timesteps += 1;
    }
    for (int j=0; j<n2; j++){
        right[j] = (dt[mid+1+j]);
        no_timesteps += 1;
    }

    int i=0;
    int j=0;
    int k = lower;

    while (i<n1 && j<n2){
        if (left[i]<=right[j]){
            dt[k] = left[i];
            k+=1;
            i+=1;
        }
        else{
            dt[k] = right[j];
            k+=1;
            j+=1;
        }
        no_timesteps += 1;
    }

    while (i<n1){
        dt[k]=left[i];
        k+=1;
        i+=1;
        no_timesteps += 1;
    }

    while (j<n2){
        dt[k] = right[j];
        k+=1;
        j+=1;
        no_timesteps += 1;
    }
}

void Sorting::quick(){
    no_recursive_calls = 0;
    no_timesteps = 0;
    no_swaps = 0;
    quick_sort(data, 0, n-1);
}

void Sorting::quick_sort(std::vector<int>& a, int lower, int upper){
    no_recursive_calls += 1;
    if (lower<upper){
        int pivot = a[upper];
        int i = lower-1;
        for (int j=lower; j<=upper-1; j++){
            no_timesteps += 1;
            if (a[j]<=pivot){
                i += 1;
                std::swap(a[j],a[i]);
                no_swaps += 1;
            }
        }
        std::swap(a[i+1],a[upper]);
        no_swaps += 1;
        quick_sort(a,lower, i);
        quick_sort(a, i+2, upper);
    }
}

void Sorting::setData(std::vector<int> a){
    data = a;
}

std::vector<int> Sorting::getData(){
    return data;
}

int Sorting::getTimeSteps(){
    return no_timesteps;
}

int Sorting::getRecursiveCalls(){
    return no_recursive_calls;
}

int Sorting::getSwaps() {
    return no_swaps;
}

File: labneura/algorithms/sorting.py
import algocpp
class Sorting:
    def __init__(self, data) -> None:
        self.data = data
        self.size = len(self.data)
        self.no_timesteps = 0
        self.no_swaps = 0
        self.no_recursive_calls = 0
        self.cpp_sorting = algocpp.Sorting(self.data)

    def selection_cpp(self):
        self.cpp_sorting.selection()
        self.data = self.cpp_sorting.data
        self.no_timesteps = self.cpp_sorting.no_timesteps
        self.no_recursive_calls = self.cpp_sorting.no_recursive_calls
        self.no_swaps = self.cpp_sorting.no_swaps

    def selection(self):
        """
        (n-1)*(n-2)*...1 = n(n-1)/2 = O(n^2)
        Irrespective of the input.
        Max no of swaps n-1 -> Max element at first and remaining all are sorted.
        Min no of swaps 0 -> Already sorted.
        """
        self.no_swaps = 0
        self.no_timesteps = 0
        for i in range(self.size-1):
            minIdx = i
            for j in range(i+1,self.size):
                if self.data[minIdx] > self.data[j]:
                    minIdx = j
                self.no_timesteps += 1
            if minIdx != i:
                self.no_swaps += 1
                (self.data[minIdx],self.data[i]) = (self.data[i],self.data[minIdx])

    def bubble_cpp(self):
        self.cpp_sorting.bubble()
        self.data = self.cpp_sorting.data
        self.no_timesteps = self.cpp_sorting.no_timesteps
        self.no_recursive_calls = self.cpp_sorting.no_recursive_calls
        self.no_swaps = self.cpp_sorting.no_swaps

    def bubble(self):
        """
        (n-1)+(n-2)+....1 = n(n-1)/2 = O(n^2)
        Sorted list -> (n-1) = O(n) = 0 Swaps
        Worst Case -> Descending Order -> O(n^2) = (n-1)+(n-2)+..+1 = n(n-1)/2 Swaps
        """
        self.no_swaps = 0
        self.no_timesteps = 0
        for i in range(self.size-1):
            swapped = False
            for j in range(self.size-i-1):
                self.no_timesteps += 1
                if self.data[j]>self.data[j+1]:
                    (self.data[j],self.data[j+1]) = (self.data[j+1],self.data[j])
                    swapped = True
                    self.no_swaps += 1
            if not swapped:
                break

    def insertion_cpp(self):
        self.cpp_sorting.insertion()
        self.data = self.cpp_sorting.data
        self.no_timesteps = self.cpp_sorting.no_timesteps
        self.no_recursive_calls = self.cpp_sorting.no_recursive_calls
        self.no_swaps = self.cpp_sorting.no_swaps

    def insertion(self):
        """
        1 + 2 + .. + (n-1) = n(n-1)/2
        Sorted List -> n = 0 insertions
        Descending List -> (n(n-1)/2) no of insertions
        """
        self.no_timesteps = 0
        for i in range(1,self.size):
            mini = self.data[i]
            j = i-1
            self.no_timesteps += 1
            inside = False
            while (j>=0 and self.data[j]>mini):
                inside = True
                self.data[j+1] = self.data[j]
                j = j-1
                self.no_timesteps += 1
            if inside:
                self.no_timesteps -= 1
            self.data[j+1] = mini

    def merge_cpp(self):
        self.cpp_sorting.merge()
        self.data = self.cpp_sorting.data
        self.no_timesteps = self.cpp_sorting.no_timesteps
        self.no_recursive_calls = self.cpp_sorting.no_recursive_calls
        self.no_swaps = self.cpp_sorting.no_swaps

    def merge(self):
        self.no_recursive_calls = 0
        self.no_timesteps = 0
        self.merge_sort(self.data)

    def merge_sort(self, arr):
        """
        Stack trace -> logn -> memory
        Each stack -> n1+n2 -> extra memory (left, and right)
        Time -> nlogn
        """
        self.no_recursive_calls += 1
        if len(arr) > 1:
            mid = len(arr) // 2
            left = arr[:mid]
            right = arr[mid:]
            self.no_timesteps += len(arr)
            self.merge_sort(left)
            self.merge_sort(right)

            i = j = k = 0

            while i<len(left) and j<len(right):
                if left[i] < right[j]:
                    arr[k] = left[i]
                    i+=1
                    k+=1
                else:
                    arr[k] = right[j]
                    j+=1
                    k+=1
                self.no_timesteps += 1
            while i<len(left):
                arr[k] = left[i]
                i+=1
                k+=1
                self.no_timesteps += 1
            while j<len(right):
                arr[k] = right[j]
                j+=1
                k+=1
                self.no_timesteps += 1

    def quick_cpp(self):
        self.cpp_sorting.quick()
        self.data = self.cpp_sorting.data
        self.no_timesteps = self.cpp_sorting.no_timesteps
        self.no_recursive_calls = self.cpp_sorting.no_recursive_calls
        self.no_swaps = self.cpp_sorting.no_swaps

    def quick(self):
        self.no_timesteps = 0
        self.no_recursive_calls = 0
        self.quick_sort(self.data, 0, len(self.data)-1)

    def quick_sort(self, arr, lower, upper):
        self.no_recursive_calls += 1
        if lower < upper:
            pivot = arr[upper]
            i = lower-1
            for j in range(lower,upper):
                self.no_timesteps += 1
                if arr[j]<pivot:
                    i+=1
                    (arr[i],arr[j]) = (arr[j], arr[i]) 
            (arr[i+1],arr[upper]) = (arr[upper], arr[i+1])
            self.quick_sort(arr, lower, i)
            self.quick_sort(arr, i+2, upper)